onnxruntime/core/optimizer/skip_layer_norm_fusion.cc

```diff
@@ -23,76 +23,209 @@ static bool IsSupportedDataType(const Node& node) {
   return true;
 }
 
+static bool CheckFirstAdd(Node& add, ProviderType providertype) {
+  if (providertype != add.GetExecutionProviderType() ||
+      !IsSupportedDataType(add)) {
+    return false;
+  }
+
+  // Check the input dimensions of the "Add" node.
+  const TensorShapeProto* add_input1_shape = add.MutableInputDefs()[0]->Shape();
+  const TensorShapeProto* add_input2_shape = add.MutableInputDefs()[1]->Shape();
+
+  if (add_input1_shape == nullptr || add_input2_shape == nullptr) {
+    return false;
+  }
+  // "Add" inputs have to be 3d.
+  if (add_input1_shape->dim_size() != 3 || add_input2_shape->dim_size() != 3) {
+    return false;
+  }
+  // "Add" inputs have to be of same dimensions.
+  bool is_valid_input = true;
+  for (int i = 0; i < 3; i++) {
+    if (add_input1_shape->dim(i).dim_value() != add_input2_shape->dim(i).dim_value()) {
+      is_valid_input = false;
+      break;
+    }
+  }
+  return is_valid_input;
+}
+
+// Add2 is the 2nd add of the to be fused sub-graph
+// The 1st input should be a 3D tensor
+// The 2nd input should be a 1D constant value
+static bool CheckSecondAdd(Node& add, ProviderType providertype) {
+  if (providertype != add.GetExecutionProviderType() ||
+      !IsSupportedDataType(add)) {
+    return false;
+  }
+
+  // Check the input dimensions of the "Add" node.
+  const TensorShapeProto* add_input1_shape = add.MutableInputDefs()[0]->Shape();
+  const TensorShapeProto* add_input2_shape = add.MutableInputDefs()[1]->Shape();
+
+  if (add_input1_shape == nullptr || add_input2_shape == nullptr) {
+    return false;
+  }
+
+  return add_input1_shape->dim_size() == 3 && add_input2_shape->dim_size() == 1;
+}
+
 /**
-Skip Layer Normalization will fuse Add + LayerNormalization into one node.
+Skip Layer Normalization will fuse Add + LayerNormalization into one node, and another Add if applicable
+
+Before fusion:
+Format 1:
+      [Sub1]   [Sub2]
+         \       /
+        Add2    /
+           \   /
+            Add1
+             |
+     LayerNormalization
+
+Format 2:
+      [Sub1]   [Sub2]
+         \       /
+          \    Add2
+           \   /
+            Add1
+             |
+     LayerNormalization
+
+Format 3:
+      [Sub1]   [Sub2]
+         \       /
+          \     /
+           \   /
+            Add1
+             |
+     LayerNormalization
+
+After fusion:
+       [Sub1]   [Sub1]
+         \      /
+          \    /
+    SkipLayerNormalization
+
+Note: This fusion doesn't consider the following case:
```

Comment:
I think this case will match with either format 1 or 2.

Reply:
if format 1 or 2 is matched, we won't consider format 3

Reply:
I mean the following case:

Note: This fusion doesn't consider the following case:
      [Sub1]   [Sub2]
         \       /
        Add2  Add3
           \   /
            Add1
             |
     LayerNormalization

It will match with Format 1 if add2 has bias, or Format2 if add3 has bias.